File: problem_12/problem_12.cpp
#include <iostream>
#include <cmath>
#include <set>

using namespace std;

int triangle_number(int before, int n) {
    return before + n;
}

int factors(int n) {
    int step = n % 2 ? 2 : 1;
    set<int> facs;
    for (int i = 1; i <= sqrt(n); i += step) {
        if (n % i == 0) {
            facs.insert(i);
            facs.insert(int(n / i));
        }
    }
    return facs.size();
}

int main() {
    int before = 0, n = 1;
    while (true) {
        int tn = triangle_number(before, n);
        cout << tn << endl;
        if (factors(tn) > 500) {
            cout << "Result: " << tn << endl;
            break;
        }
        before = tn;
        n++;
    }
    return 0;
}

File: problem_12/problem_12.rs
use std::collections::HashMap;


fn triangle_number(before: u32, n: u32) -> u32 {
    before + n
}

fn factors(n: u32) -> u32 {
    let mut step = 2;
    if n % 2 == 0 {
        step = 1;
    }
    let mut facs = HashMap::new();
    for i in (1..((n as f64).sqrt() as u32) + 1).step_by(step) {
        if n % i == 0 {
            facs.insert(i, true);
            facs.insert(n / i, true);
        }
    }
    facs.len() as u32
}

fn main() {
    let mut n: u32 = 1;
    let mut before: u32 = 0;
    loop {
        let tn = triangle_number(before, n);
        println!("{}", tn);
        if factors(tn) > 500 {
            println!("Result: {}", tn);
            break;
        }
        before = tn;
        n += 1;
    }
}
